<a href="https://colab.research.google.com/github/TheCiceerGuy/Decision-Tree-for-Titanic-Dataset/blob/main/DecisionTree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing Numpy and Pandas**

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing


**Importing Libraries**

In [3]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

**Load the dataset**

In [8]:
train_data = pd.read_csv('/content/train.csv')

**Check the Dataset**

In [27]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0      1      0   
2                             Heikkinen, Miss. Laina    1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0      1      0   
4                           Allen, Mr. William Henry    0  35.0      0      0   

             Ticket     Fare Cabin  Embarked  
0         A/5 21171   7.2500   NaN       0.0  
1          PC 17599  71.2833   C85       1.0  
2  STON/O2. 3101282   7.9250   NaN       0.0  
3            113803  53.1000  C123       0.0  
4            373450   8.0500   NaN       0.0

**Handle missing values**

In [9]:
train_data['Age'].fillna(train_data['Age'].median(), inplace=True)

Encode categorical variables like 'Sex' and 'Embarked'

In [10]:
train_data['Sex'] = train_data['Sex'].map({'male': 0, 'female': 1})
train_data['Embarked'] = train_data['Embarked'].map({'S': 0, 'C': 1, 'Q': 2})

**Define features (X) and target variable (y)**

In [11]:
X = train_data[['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']]
y = train_data['Survived']

**Split the data into training and testing sets**

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

**Implement Decision Tree Classifier:**

In [13]:
clf = DecisionTreeClassifier()

**Train the classifier**

Here we handle missing values in your dataset b using Imputation method.

In [15]:
from sklearn.impute import SimpleImputer

In [16]:
imputer = SimpleImputer(strategy='median')

In [17]:
imputer.fit(X_train)

SimpleImputer(strategy='median')

In [19]:
X_train_imputed = imputer.transform(X_train)

In [21]:
X_test_imputed = imputer.transform(X_test)

In [23]:
clf.fit(X_train_imputed, y_train)

DecisionTreeClassifier()

**Evaluate the Model: Make Predictions on the set**

In [24]:
y_pred = clf.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:432: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(


**Calculate Accuracy**

In [28]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 78.77%


After running, we obtain the accuracy of the Decision Tree classifier in predicting survival on the Titanic.

#Analysis of Titanic Dataset using Decision Trees:



1.   Model Performance:      

*   Accuracy: The model shwd an accuracy of 78.77%.




2.   Insights and Observations


*   Key Features: Sex, Age, Embarked, Fare were the most influential features affecting survival predictions.
*   Model Limitations: Moodel might not perform well if data is of non-related classes.





#Sample Summary:

Summary of Results:

1.   The Decision Tree model achieved an accuracy of 78.77% on the test set.
2.   The confusion matrix shows that the model predicts survived vs. not survived with high/low accuracy.



Insights and Observations:

1.   Age and sex were among the most important factors affecting survival, with females and younger individuals having higher chances of survival.
2.  The model performs well in predicting survival based on [age, sex, embarked], but it struggles with [Sibsp, Fare].



**Conclusion:**

 The Decision Tree model provides valuable insights into survival prediction on the Titanic dataset, with notable importance placed on certain features. However, further enhancements and explorations could refine the model's accuracy and robustness.